File: app/src/main/java/com/example/roomscanpro/data/InMemoryRoomRepository.kt
package com.example.roomscanpro.data

import java.util.UUID

object InMemoryRoomRepository {
    private val rooms: MutableList<Room> = mutableListOf(
        Room(UUID.randomUUID().toString(), "Living Room", "Living Room", 5.0, 4.0, 2.7),
        Room(UUID.randomUUID().toString(), "Bedroom", "Bedroom", 4.0, 3.2, 2.7)
    )

    fun listRooms(): List<Room> = rooms.toList()

    fun createRoom(
        name: String,
        type: String,
        length: Double,
        width: Double,
        height: Double
    ): Room {
        val newRoom = Room(UUID.randomUUID().toString(), name, type, length, width, height)
        rooms.add(newRoom)
        return newRoom
    }

    fun findRoom(roomId: String): Room? = rooms.find { it.roomId == roomId }

    fun updateRoom(
        roomId: String,
        name: String,
        type: String,
        length: Double,
        width: Double,
        height: Double
    ): Boolean {
        val room = findRoom(roomId) ?: return false
        room.roomName = name
        room.roomType = type
        room.lengthMeters = length
        room.widthMeters = width
        room.heightMeters = height
        return true
    }

    fun deleteRoom(roomId: String): Boolean {
        return rooms.removeIf { it.roomId == roomId }
    }
}

File: flutter_app/lib/data/in_memory_room_repository.dart
import 'dart:math';
import 'room.dart';
import 'furniture.dart';

class InMemoryRoomRepository {
  static final InMemoryRoomRepository _instance = InMemoryRoomRepository._internal();

  factory InMemoryRoomRepository() {
    return _instance;
  }

  InMemoryRoomRepository._internal() {
    // Seed some data
    var livingRoom = Room(id: _generateUuid(), name: 'Living Room', type: 'Living Room', length: 5.0, width: 4.0, height: 2.7);
    livingRoom.furniture.add(Furniture(id: _generateUuid(), name: 'Sofa', length: 2.0, width: 0.8, height: 1.0, color: 'Grey'));
    _rooms.add(livingRoom);

    _rooms.add(Room(id: _generateUuid(), name: 'Bedroom', type: 'Bedroom', length: 4.0, width: 3.2, height: 2.7));
  }

  final List<Room> _rooms = [];

  List<Room> listRooms() => _rooms;

  Room createRoom(String name, String type, double length, double width, double height) {
    final newRoom = Room(id: _generateUuid(), name: name, type: type, length: length, width: width, height: height);
    _rooms.add(newRoom);
    return newRoom;
  }

  Room? findRoom(String id) {
    try {
      return _rooms.firstWhere((room) => room.id == id);
    } catch (e) {
      return null;
    }
  }

  bool updateRoom(String id, String name, String type, double length, double width, double height) {
    final room = findRoom(id);
    if (room == null) return false;
    room.name = name;
    room.type = type;
    room.length = length;
    room.width = width;
    room.height = height;
    return true;
  }

  bool deleteRoom(String id) {
    final room = findRoom(id);
    if (room == null) return false;
    return _rooms.remove(room);
  }

  // --- Furniture Operations ---

  void addFurniture(String roomId, String name, double length, double width, double height, String color) {
    final room = findRoom(roomId);
    if (room != null) {
      room.furniture.add(Furniture(
        id: _generateUuid(),
        name: name,
        length: length,
        width: width,
        height: height,
        color: color,
      ));
    }
  }

  bool deleteFurniture(String roomId, String furnitureId) {
    final room = findRoom(roomId);
    if (room != null) {
      room.furniture.removeWhere((f) => f.id == furnitureId);
      return true;
    }
    return false;
  }

  String _generateUuid() {
    return List.generate(16, (index) => Random().nextInt(256))
        .map((i) => i.toRadixString(16).padLeft(2, '0'))
        .join();
  }
}